### PR TITLE
Add `topological_sort()` to `datanames(teal_data())` and also extend `datanames()` with parent dataset when it is provided in `join_keys`

### PR DESCRIPTION
**Pull request by m7pr**
Part of https://github.com/insightsengineering/teal/pull/1253

This PR introduced below changes
- `teal_data()` constructor does not put `names(join_keys)` in default `datanames(teal_data())` to maintain consistency with other features, where we do not allow `datanames()` to contain names of objects that do not exist in `@env`
- `datanames()` now sorts names topologically based on provided `join_keys()`
- each time `datanames()` and `join_keys()` is overwritten the sort is applied to `teal_data()@datanames`
- provided few more tests
- adjusted 2 tests that assumed extraction of `datanames()` from `join_keys` is fine - but it's not right now as we do not allow `datanames()` to contain names of objects that do not exists in `@env`

Will provide testing in this PR in teal as well https://github.com/insightsengineering/teal/pull/1280/files

### CODE REVIEW

**Comment by m7pr** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by gogonzo** [comment]
@m7pr please add a NEWS entry and fix GHA checks

**Comment by m7pr** [comment]
Hey @gogonzo NEWS enhanced! that was a long one.
Fixed lintr and spelling checks as well!

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,15 @@
 # teal.data 0.6.0.9005
 
+### Enhancements
+
+- `datanames()`
```
Comment (gogonzo) [inline NEWS.md:5]:
```suggestion
- change the way of setting default `datanames()`:
```

**Comment by gogonzo** [comment]
👍

**Review by gogonzo — APPROVED** [review]

**Pull request merged**